### PR TITLE
Added manager for OS service

### PR DESCRIPTION
**Pull request by ericm**
Related issue: #107 
The manager will hold the state of the OS's installed for the OS service.

### CODE REVIEW

File: gnoi/os/manager.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		return nil
+	}
+	if supported := m.supportedVersions[o.Version]; !supported {
+		return fmt.Errorf("INCOMPATIBLE")
+	}
+	m.osMap[o.Version] = o
+	return nil
+}
+
+// Rollback to factory OS. Wipes all other OS's.
```
Comment (loriscn12) [inline gnoi/os/manager.go:90]: Rollbacks ... and wipes...

File: gnoi/os/manager.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+// SetRunning OS to new version. OS must exist in manager.
+func (m *Manager) SetRunning(version string) error {
+	if o, ok := m.osMap[version]; ok && o != nil {
+		m.running = o
+	}
+	return fmt.Errorf("NON_EXISTENT_VERSION")
+}
+
+// Install an OS. Must be fully transfered and verified beforehand.
```
Comment (loriscn12) [inline gnoi/os/manager.go:78]: Installs

File: gnoi/os/manager.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+func (m *Manager) Install(o *OS) error {
+	if _, ok := m.osMap[o.Version]; ok {
+		return nil
+	}
+	if supported := m.supportedVersions[o.Version]; !supported {
+		return fmt.Errorf("INCOMPATIBLE")
+	}
+	m.osMap[o.Version] = o
+	return nil
```
Comment (loriscn12) [inline gnoi/os/manager.go:87]:
```suggestion
	if supported := m.supportedVersions[o.Version]; !supported {
		return fmt.Errorf("INCOMPATIBLE")
	}
	if _, ok := m.osMap[o.Version]; !ok {
		m.osMap[o.Version] = o
	}
	return nil
```

File: gnoi/os/manager.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+// IsRunning the version of OS specified.
+func (m *Manager) IsRunning(version string) bool {
+	if m.running != nil {
+		return m.running.Version == version
+	}
+	return false
+}
+
+// SetRunning OS to new version. OS must exist in manager.
```
Comment (loriscn12) [inline gnoi/os/manager.go:70]: Sets the running OS...

File: gnoi/os/manager.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+func (m *Manager) SetRunning(version string) error {
+	if o, ok := m.osMap[version]; ok && o != nil {
```
Comment (loriscn12) [inline gnoi/os/manager.go:72]: can the entry be nil when ok?

**Comment by ericm** [comment]
The transfer state is all going to be handled in the OS Server. This will just manage the OS's after the fact

File: gnoi/os/manager.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+func NewManager(supported map[string]bool, factoryOs *OS) *Manager {
+	return &Manager{
+		standbyState:      unsupported,
+		supportedVersions: supported,
+		factory:           factoryOs,
+		osMap:             map[string]*OS{factoryOs.Version: factoryOs},
+	}
+}
+
+// IsRunning the version of OS specified.
```
Comment (loriscn12) [inline gnoi/os/manager.go:62]: drop it or make it more descriptive

File: gnoi/os/manager.go (hunk 5 above)
Comment (ericm) [reply gnoi/os/manager.go:72]: It can be but in normal circumstances I suppose it's redundant

File: gnoi/os/manager.go (hunk 2 above)
Comment (ericm) [reply gnoi/os/manager.go:78]: Does the first word of the comment not have to be the method name?

Comment (loriscn12) [reply gnoi/os/manager.go:78]: Yep leave that , you can have:

Install installs an OS ...

File: gnoi/os/manager.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// OS represents the data for a singular operating system.
+type OS struct {
+	Name,
+	Hash,
+	Version string
+	BuildDate time.Time
+	Data      *bytes.Buffer
+}
+
+// Manager for storing data on OS's.
```
Comment (samribeiro) [inline gnoi/os/manager.go:33]: I like how you have set this up has if we are running real OSs in the Target =)

I think we can simplify here a bit (this is what I think, I may be wrong, please point out if so):
1) Manager only really needs: runningOS, factoryOS and []installeOSs string. We dont do anything with all the other data.
2) The hash validation is only done at the reception of the binary data, so we dont need to store the hashes;
3) Manager should support all versions, we can signal an unsupported version in the OS file directly by using a cookie, example:

OS file:
hash (for everything after this line)\
cookie (just a unique id string) \
version (string) \
random padding \
<EOF>

Comment (ericm) [reply gnoi/os/manager.go:33]: So for now just keep OS here and later move it?

**Comment by ericm** [comment]
I simplified it according to your suggestions :)

File: gnoi/os/manager.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	return version == m.runningVersion
+}
+
+// SetRunning sets the running OS to the version specified.
+func (m *Manager) SetRunning(version string) error {
+	if _, ok := m.osMap[version]; ok {
+		m.runningVersion = version
+		return nil
+	}
+	return fmt.Errorf("NON_EXISTENT_VERSION")
```
Comment (samribeiro) [inline gnoi/os/manager.go:48]: we may need to revisit this after but good for now.

**Pull request merged**